source/common/router/rds_impl.cc

```diff
@@ -73,6 +73,24 @@ RdsRouteConfigSubscription::RdsRouteConfigSubscription(
       factory_context.clusterManager(), factory_context.random(), *scope_,
       "envoy.api.v2.RouteDiscoveryService.FetchRoutes",
       "envoy.api.v2.RouteDiscoveryService.StreamRoutes", factory_context.api());
+
+  initialization_timeout_ = std::chrono::milliseconds(
+      PROTOBUF_GET_MS_OR_DEFAULT(rds.config_source(), initial_fetch_timeout, 0));
+  if (initialization_timeout_.count() > 0) {
+    initialization_timeout_timer_ = factory_context.dispatcher().createTimer([this]() -> void {
+      ENVOY_LOG(warn, "rds: initialization timed out for route_config_name={}", route_config_name_);
+      onConfigUpdateFailed(nullptr);
+    });
+  }
+}
+
+void RdsRouteConfigSubscription::initialize(std::function<void()> callback) {
+  initialize_callback_ = callback;
+  if (initialization_timeout_.count() > 0) {
+    ASSERT(initialization_timeout_timer_);
+    initialization_timeout_timer_->enableTimer(initialization_timeout_);
+  }
+  subscription_->start({route_config_name_}, *this);
```

Comment:
Do you think we could put the timeout inside the `subscription_->start()`? That way we only need to implement once for gRPC and once for REST.

Reply:
I am trying to implement this. 
Do you think, in that case, we should move `initial_fetch_timeout` parameter to `config_source.api_config_source` or leave it in `config_source`? 

If we leave it on `config_source` level, it could be set for "path" config source type, but we don't want that - it will be not implemented for that type. Maybe it is not a problem if we mention that in docs?

If we move it to `api_config_source`, there will be a problem with ADS, because i believe we do not have access to ads mux config in [subscription_factory.h](/envoyproxy/envoy/blob/master/source/common/config/subscription_factory.h#L87). Moreover LDS, RDS and CDS would have to share the same timeout.

Reply:
I think it's fine to leave this on `config_source`. While it won't make sense for all source types, we can add explicit checks to reject or just ignore. Also, even for file based options, timeouts might make sense, since they use inotify and watch for updates.

Reply:
Done, I moved timeout to Subscription.

I also looked closer at [filesystem_subscription_impl.h](https://github.com/envoyproxy/envoy/blob/master/source/common/config/filesystem_subscription_impl.h) and I think (correct me if I'm wrong) it doesn't make much sense to implement the timeout there. I think filesystem subscription always calls `onConfigUpdate`/`onConfigUpdateFailed` immediately after it starts.

Reply:
Sure, this is fine for now, it's conceivable that we would want to timeout on inotify, but I doubt anyone would use it today.